lib/pbench/server/api/resources/pbench_users.py

```diff
@@ -0,0 +1,440 @@
+                    "auth_token": auth_token.decode(), # Will not present if failed
+                }
+        """
+        # get the post data
+        post_data = request.get_json()
+        if not post_data:
+            self.logger.warning("Invalid json object: {}", request.url)
+            abort(400, message="Invalid json object in request")
+
+        username = post_data.get("username")
+        if not username:
+            self.logger.warning("Username not provided during the login process")
+            abort(400, message="Please provide a valid username")
+
+        password = post_data.get("password")
+        if not password:
+            self.logger.warning("Password not provided during the login process")
+            abort(400, message="Please provide a valid password")
+
+        try:
+            # fetch the user data
+            user = (
+                Database.db_session.query(UserModel)
+                .filter_by(username=username)
+                .first()
+            )
+        except Exception:
+            self.logger.exception("Exception occurred during user login")
+            abort(500, message="INTERNAL ERROR")
+
+        if not user:
+            self.logger.warning(
+                "No user found in the db for Username: {} while login", username
+            )
+            abort(403, message="No such user, please register first")
+
+        # Validate the password
+        if not check_password_hash(user.password, password):
+            self.logger.warning("Wrong password for user: {} while login", username)
+            abort(401, message="Wrong password")
+
+        try:
+            auth_token = self.auth.encode_auth_token(
+                self.token_expire_duration, user.id
+            )
+        except (
+            jwt.InvalidIssuer,
+            jwt.InvalidIssuedAtError,
+            jwt.InvalidAlgorithmError,
+            jwt.PyJWTError,
+        ):
+            self.logger.error(
+                "Could not encode the JWT auth token for user: {} while login", username
+            )
+            abort(
+                500, message="INTERNAL ERROR",
+            )
+
+        # check if the user is already logged in
+        # We would still proceed to login and return a new auth token to the user
+        if (
+            self.auth.token_auth.current_user()
+            and user == self.auth.token_auth.current_user()
+        ):
+            self.logger.warning("User already logged in and trying to re-login")
+            if self.auth.token_auth.get_auth()["token"] == auth_token.decode():
+                # If a user attempts to re-login immediately within a second,
+                # the new auth token will be same as the one present in the header
+                # Therefore we will not fulfill the re-login request
+                self.logger.warning("Attempted immediate re-login")
+                abort(403, message="Retry login after some time")
```

Comment:
This could be an issue for testing; but for real life, some sort of login rate-limit is a great idea anyway.

Reply:
Can you explain more about how this works?  I am not following this code.  When `encode_auto_token()` is called, won't it generate a different token each time because it is based on time?  The `datetime` object I believe has sub-second resolution.

Reply:
This was exactly an issue I found while testing, that if I tried to create another token immediately after creating one token (re-login on the very next line of the code) the tokens are exact similar. Peter, I think I should I update the comment of re-login within 1 second, it should be within maybe 1 millisecond (or any delta small enough to cause the same auth token)

Reply:
Perhaps add a monotonically incrementing token ID counter?

Reply:
The token is auto generated using a jwt package and it is based on the current utc time. Adding a token Id counter might be a good idea but there isnt a direct way to do that. I will create an issue to track this in a separate PR.